File: src/main/java/net/moddingplayground/onastick/api/screen/LocalScreenHandlerFactory.java
package net.moddingplayground.onastick.api.screen;

import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.item.Item;
import net.minecraft.screen.ScreenHandler;
import net.minecraft.screen.ScreenHandlerContext;
import net.minecraft.screen.ScreenHandlerFactory;

import java.util.function.Supplier;

@FunctionalInterface
public interface LocalScreenHandlerFactory {
    ScreenHandler create(int syncId, PlayerInventory inventory, ScreenHandlerContext context, Item item);

    static ScreenHandlerFactory build(LocalScreenHandlerFactory factory, Supplier<Item> item) {
        return (syncId, inventory, player) -> factory.create(syncId, inventory, ScreenHandlerContext.create(player.getWorld(), player.getBlockPos()), item.get());
    }
}

File: src/main/kotlin/dev/andante/onastick/screen/LocalScreenHandlerFactory.kt
package dev.andante.onastick.screen

import java.util.function.Supplier
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.item.Item
import net.minecraft.screen.ScreenHandler
import net.minecraft.screen.ScreenHandlerContext
import net.minecraft.screen.ScreenHandlerFactory

/**
 * A screen factory for a local handled screen.
 */
fun interface LocalScreenHandlerFactory {
    fun create(syncId: Int, inventory: PlayerInventory, context: ScreenHandlerContext, item: Item): ScreenHandler

    companion object {
        fun build(factory: LocalScreenHandlerFactory, item: Supplier<Item>): ScreenHandlerFactory {
            return ScreenHandlerFactory { syncId, inventory, player ->
                factory.create(
                    syncId,
                    inventory,
                    ScreenHandlerContext.create(player.world, player.blockPos),
                    item.get()
                )
            }
        }
    }
}
